### ISSUE TITLE
Key hierarchies and combinations

### ISSUE THREAD

**Issue by mattsb42-aws** [awslabs/aws-encryption-sdk-specification#45]
Creating this for a space to discuss a feature request: https://github.com/aws/aws-encryption-sdk-python/issues/201

The request is for the ability to define a key hierarchy that would allow intermediate key selection based on encryption context, potentially combining multiple intermediate keys, with the end goal of keeping the root of trust in a single master key but enabling cryptographic deletion of data with specific encryption contexts.

Thinking about this, I think the described behavior breaks down into a few necessary features that would create separate layers to this:

1. Tooling to define a key hierarchy and store intermediate keys.
1. A key selection mechanism to provide extension points for how to select the correct key.
1. A mechanism to select more than one key and combine them.

I think that layer 1 alone would be useful, as would layers 1+2, in addition to 1+2+3 doing what you described.

## Key Hierarchy

In addition to use cases like that described in the original feature request, I think that tooling to manage a key hierarchy could also provide a nice alternative to data key caching that would still allow a reduction in call frequency to an upstream keyring. I'm envisioning something, probably a CMM, that behaves in a similar way to the KMS re-authorization flow in DynamoDB server-side encryption[1]. This would enable a reduction in KMS call frequency while still enabling each encrypted message to have a unique data key.

A naive solution to this could be to simply have the encrypted data key be an AWS Encryption SDK encrypted message, but I do not think that this would scale well, especially once we introduce the concept of multiple intermediate keys, either as separate layers or as required combinations.

## Intermediate Key Selection

A key selection extension point would enable users to define custom criteria for locating an intermediate key. This could be based on the encryption context, the keyring trace, or anything else in the encryption/decryption materials request.

## Multiple Intermediate Keys

I really like the idea outlined in the original feature request of a single encrypted message requiring multiple, independent, intermediate keys.

There are two basic ways that I could see these multiple keys being used: either nested within each other with a chain of key hierarchy keyrings, or in parallel with a single keyring having intermediate key selection criteria that can result in multiple keys and those keys being combined in some way (HKDF? XOR?) to arrive at the actual intermediate key material.

The parallel approach has some benefits over the nested approach, such as reduced encrypted data key size and simpler configuration.

I think that "both" is probably the right answer here.

[1] https://docs.aws.amazon.com/kms/latest/developerguide/services-dynamodb.html#dynamodb-encrypt

**Comment by mattsb42-aws** [comment]
@ikben could you confirm whether I correctly captured your use cases above?

**Comment by mattsb42-aws** [comment]
Discussing this with @SalusaSecondus today, he brought up an alternate solution to the original request.

Rather than building a key hierarchy and using the custom value as a key, the custom value could be used as a mix-in for a KDF.

This would avoid the need to support rotation of those values, but would still provide the original functionality of scoped cryptographic deletion.

Critically, though, one thing that this would *not* change is the sensitivity of the custom values. They would still need to be protected with the same level of protections as if they were the encryption keys themselves, and so would need most of the infrastructure required for proposed solution **1.1 Key Hierarchy**.

**Comment by SalusaSecondus** [comment]
Yes. Using a KDF to combine a per-domain (e.g., user) secret with normally managed data keys does mean we don't need to support rotation of the per-domain secret. If we just did wrapping then rotation would be necessary to prevent cryptographic wear-out.

Selection of these secrets is still TBD as is their protection (as noted by @mattsb42-aws ). One important thing is that these secrets are (by necessity) _not_ stored as part of the ciphertext. This can simplify things and also makes their size less sensitive (since they are only being stored once, not once per message). Size still matters as a large number of them can be bulky and network transfer time might matter.

Possible ways to refer to these from a ciphertext include:
* Embedding their ID in the EncryptionContext. (Probably the best design.)
* Defining the concept of a "signalling encrypted data key" which cannot be used to actually acquire the data key but may change the behavior of a knowledgable CMM/Key-Right.  This is probably too complicated and overkill.

**Comment by ikben** [comment]
@mattsb42-aws Absolutely. It's great to see it summarised like that.

One thing I was recently thinking about is about how my example solution always encrypts the Datakey a) locally (exposing the domain keys of each domain to the client) and b) with the same (combined) key.

Assuming my basic understanding of Key Derivation Functions is correct, a KDF would solve (b), so I would prefer that over my naive solution of xor'ing.
